[WIP] Sequencing data

LogvinovLeon opened the pull request:
# Description

TODO

-------


- [ ] I have updated `CHANGELOG.md` with a new entry if my PR makes any breaking changes or fixes a bug. If my PR removes a feature or changes its behavior, I provide help for users on how to migrate to the new behavior.
- [ ] I have carefully reviewed all my `Cargo.toml` changes before opening the PRs. (Are all new dependencies necessary? Is any module dependency leaked into the full-node (hint: it shouldn't)?)

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

Rendered docs are available at `https://sovlabs-ci-rustdoc-artifacts.us-east-1.amazonaws.com/<BRANCH_NAME>/index.html`

LogvinovLeon committed:
Split transaction module into unsigned.rs, v0.rs and v1.rs

LogvinovLeon committed:
Remove redundant custom PartialEq,Eq implementations

LogvinovLeon committed:
Move PubKeyAndSignature into V1

bkolad commented on crates/module-system/sov-modules-api/src/transaction/types/v2.rs:
```diff
@@ -0,0 +1,72 @@
+    /// The transaction metadata. Contains gas parameters and the chain ID.
+    pub details: TxDetails<S>,
+    /// Sequencer metadata. Can contain things like high-precision timestamps
+    pub sequencing_data: Vec<u8>,
```
What’s the reasoning behind creating a new transaction type instead of adding this functionality to `V0` and `V1`? If we need to do something similar for transactions that support multi-sig, we would need to introduce a V4. Is the concern simply that if some V0 and V1 txs don’t use this field, they would waste space storing an empty Vec?

LogvinovLeon replied:
My reasoning was that it would be a breaking change and people would not be able to deserialize txs they already have in the database. If we're ok with that - sure. But then by this logic - shouldn't we also remove V0 and V1 and just have V0 with multisig that can have only 1 signature?

bkolad reviewed (commented):
This PR LGTM and thx for the code cleanup, which improved the code structure. 
I had only one question

LogvinovLeon committed:
Simplify auth

LogvinovLeon committed:
Add sequencing data to context

LogvinovLeon committed:
Add database persistence for sequencing_data

- Add sequencing_data column to Postgres events table
- Update PreferredSequencerDbBackend trait to accept sequencing_data parameter
- Update both Postgres and RocksDB backends to store/retrieve sequencing_data
- Add sequencing_data_list to PreferredSequencerReadBatch structure
- Update all database call sites to handle sequencing_data tuples

The sequencing_data field stores sequencer-provided metadata (e.g., timestamps)
that is not signed by users but persists alongside transactions in the database.
This ensures data survives sequencer restarts.

LogvinovLeon committed:
WIP: Remove sequencing_data from AuthorizationData

Move sequencing_data out of AuthorizationData (it's not auth data) and add it
to PreferredBatchData structure instead. Update TransactionAuthorizer trait to
accept sequencing_data as a separate parameter.

Started updating IncrementalBatch trait to yield sequencing_data alongside
transactions. This is work in progress - need to update all implementations.

LogvinovLeon committed:
Implement sequencing_data infrastructure for transaction timestamps

Add complete end-to-end support for sequencer-provided transaction metadata:

- Add sequencing_data field to PreferredBatchData (serialized to DA)
- Add sequencing_data_list to BatchWithId and IterableBatchWithId
- Add get_sequencing_data() method to IncrementalBatch trait
- Update TransactionAuthorizer::resolve_context() to accept sequencing_data parameter
- Thread sequencing_data through process_tx_and_reward_prover() and all transaction processing
- Pre-collect sequencing_data before batch iteration to avoid borrow issues
- Populate timestamps in preferred sequencer (nanoseconds since UNIX epoch as u128)
- Pass None for standard sequencer and simulation endpoints

Data flow: Sequencer accept_tx → DB storage → DA submission → Full node replay → Batch processing → Context

Modules can now access via context.sequencing_data() to get high-precision timestamps for deterministic time-based logic.

LogvinovLeon committed:
Fix test utilities and tests for sequencing_data support

Update test utilities to initialize sequencing_data field in PreferredBatchData:
- sov-test-utils: Add empty sequencing_data vec to blob generators
- sov-blob-storage tests: Add sequencing_data to BatchWithId::new() calls
- Add Arc imports where needed for sequencing_data initialization

All unit tests passing for:
- sov-modules-api
- sov-blob-storage
- sov-modules-stf-blueprint
